### ISSUE TITLE
Exception handling

### ISSUE THREAD

**Issue by ariscn** [timothycrosley/hug#318]
Hug provides `@exception` to add error handlers. This is great, but I'd like to propose an improvement - there should be a way to exclude Falcon's error classes - these are raised by Falcon itself, or by your own code to handle *expected* problems.

Right now, I can do something like this:

    @exception()
    def handle_exception(exception):
        if not isinstance(exception, falcon.HTTPError):
            log.exception('Unexpected error while handling API request.')

        raise falcon.HTTPInternalServerError('Something broke', 'Something broke')

This works, but it would be nice to do something a bit cleaner:

    @exception(exclude_falcon=True)    # proposed
    def handle_exception(exception):
        raise falcon.HTTPInternalServerError('Something broke', 'Something broke')

In other words, a falcon exception isn't "really" an exception, and it would be useful to have a convenience function or argument to capture non-Falcon errors.

**Comment by ariscn** [comment]
Found another issue with exception handler. I'll park this here for now, let me know if you'd prefer a separate issue.

If a middleware throws an exception, it's handled by falcon and never hits hug's exception handlers.

**Comment by timothycrosley** [comment]
Hi @ariscn, 

Thanks for the ideas! I think the way I will implement the first will be an exclude=(Exception, ) tuple that can be used to define any number of class of exceptions that you would like to be handled externally. The second is known, with the assumption that middleware exceptions are different that other classes of exceptions - and should generally handle the exceptions internally in their own unique way - but I think I'm getting swayed to the other way of thinking.

Thanks!

~Timothy

**Comment by ariscn** [comment]
Similar comment to the other thread - if this is a known issue (or considered a feature at the moment), it's worth documenting.

**Comment by vortec** [comment]
An ``exclude`` argument would also prevent ``KeyboardInterruption`` exceptions being handled while developing and hitting CTRL+C constantly.

**Comment by ariscn** [comment]
Found another issue with Exception handling.

My API has an exception handler, and uses `LogMiddleware`. The error handler re-raises Falcon exceptions and raises HTTPInternalServerError(...) otherwise.

When an exception occurs, this returns a 500 as expected. But, the middleware logs a 200! Roughly:

    @middleware_class()
    class DummyLogMiddleware(LogMiddleware):
        pass


    @exception()
    def handle_exception(exception):
        raise HTTPInternalServerError('Something broke', 'Something broke')


    @get('/foo')
    def foo():
        raise Exception('foo')

**Comment by timothycrosley** [comment]
This has been officially implemented and merged into develop in preparation for the next release here: https://github.com/timothycrosley/hug/pull/382/files

Thanks!

~Timothy

**Comment by timothycrosley** [comment]
2.2.0 has now been officially pushed to PYPI with this issue resolved

Thanks!

~Timothy

**Comment by pikeas** [comment]
Quick followup question - does 2.2.0 also change handling for exceptions thrown by middleware (handled by Falcon per earlier comment), and will middleware now see the correct status code?
